### ISSUE TITLE
Smart playlists do not play any more

### ISSUE THREAD

**Issue by shoubamzlibap** [ampache/ampache#1440]
# Environment

 * Ampache version: 3.8.3 development - downloaded 06-dec-2016
 * Web server + version: nginx 1.10.2
 * Server operating system: fedora 24 (arm)
 * Client operating system: win 7 / current google chrome browser

 * Client type:
   - Web

# Settings
See attachment.

# Description

For a while smart playlists worked, then suddenly they did not show up at all any more. I checked the amapche log, and saw several corrupted tables being reported:
```
awk -F\' '/Table/ {print $2}' /var/log/ampache/ampache.log*  |sort -u
.\/ampache\/object_count
.\/ampache\/tmp_playlist_data
```

 I fixed the tables with myisamchk, like:

`myisamchk -r object_count.MYI
`

 After that, 
the playlists showed up again, but I could not play them, and the following error is displayed in the ampache log:

`2016-12-06 21:09:38 [ampache] (PHP) -> [Runtime Error] Invalid argument supplied for foreach() in file /var/www/html/ampache-develop/lib/class/search.class.php(1188)`

Regular playlists play fine.

# Logs

```
2016-12-06 10:29:11 [ampache] (Dba) -> Connection failed: SQLSTATE[HY000] [2002] No such file or directory
2016-12-06 10:29:11 [ampache] (Dba) -> Error: failed to get database handle
2016-12-06 10:29:11 [ampache] (Dba) -> Connection failed: SQLSTATE[HY000] [2002] No such file or directory
2016-12-06 10:29:52 [ampache] (PHP) -> [Runtime Error] Invalid argument supplied for foreach() in file /var/www/html/ampache-master/lib/class/search.class.php(1188)
2016-12-06 10:30:12 [ampache] (PHP) -> [Runtime Error] Invalid argument supplied for foreach() in file /var/www/html/ampache-master/lib/class/search.class.php(1188)
2016-12-06 10:30:59 [isaac] (Dba) -> Error: ["HY000",2014,"Cannot execute queries while other unbuffered queries are active.  Consider using PDOStatement::fetchAll().  Alternatively, if your code is only ever going to run against mysql, you may enable query buffering by setting the PDO::MYSQL_ATTR_USE_BUFFERED_QUERY attribute."]
2016-12-06 21:09:38 [ampache] (PHP) -> [Runtime Error] Invalid argument supplied for foreach() in file /var/www/html/ampache-develop/lib/class/search.class.php(1188)

```

You will notice that I updated to development branch for the last log line ;)

[ampache.cfg.php.txt](https://github.com/ampache/ampache/files/634950/ampache.cfg.php.txt)

**Comment by shoubamzlibap** [comment]
It seems that at least _some_ new smartplaylists play - I am having trouble with a special search, but maybe that is a different problem (maybe the number of results is too big?). My current theory: the database corruption deleted some information of the smart playlist. New smart playlists work because the database is intact again.

What seems to be missing in old smart playlists is the `$this->rules` Array. It is non-existant on non working smartplaylist, and it displays something like 

```
(
    [0] => Array
        (
            [0] => anywhere
            [1] => contain
            [2] => hans
            [3] =>
        )

)

```
on a working smart playlist. The `$this->rules` is part of the argument to the `foreach` call on line 1188, which is why I am looking at it.
I will debug further and report back.

**Comment by Phyks** [comment]
Playlist handling in database is not really resilient. And database tables corruption do happen, but this is yet to be debugged further. Any information you might have on this would be really helpful!

**Comment by shoubamzlibap** [comment]
I have attached some lines I found in the mariadb.log. It looks like there was a more severe crash, and thats is where the trouble started. The log shows only tmp_playlist_data, but object_count follows soon.
Someone more versed in mysql/mariadb might make more sense of it then myself. 
[mysql.txt](https://github.com/ampache/ampache/files/691682/mysql.txt)

The full log file is here:
[mariadb.log.txt.gz](https://github.com/ampache/ampache/files/691699/mariadb.log.txt.gz)
If you uncompress it, its 80M :)

About new smart playlists not playing: that seems to be a size problem - if the search returns a very large number of results, it can take a very long time until the playlist starts. It might not even start at all. I have not tried to distinguished these too cases. But the hardware I am using is a raspberry pi, so performance is generally rather poor. I think for the moment we can ignore this. (I have plans to try out new hardware)

Let me know what else I could do to help with this.
